codex-rs/exec-server/src/posix.rs

```diff
@@ -0,0 +1,627 @@
+                    fds.len()
+                ));
+            }
+
+            if msg
+                .fds
+                .iter()
+                .any(|src_fd| fds.iter().any(|dst_fd| dst_fd.as_raw_fd() == *src_fd))
+            {
+                return Err(anyhow::anyhow!(
+                    "overlapping fds not yet supported in SuperExecMessage"
+                ));
+            }
+
+            let mut command = Command::new(file);
+            command
+                .args(&argv[1..])
+                .arg0(argv[0].clone())
+                .envs(&env)
+                .current_dir(&workdir)
+                .stdin(Stdio::null())
+                .stdout(Stdio::null())
+                .stderr(Stdio::null());
+            unsafe {
+                command.pre_exec(move || {
+                    for (dst_fd, src_fd) in msg.fds.iter().zip(&fds) {
+                        libc::dup2(src_fd.as_raw_fd(), *dst_fd);
+                    }
+                    Ok(())
+                });
+            }
+            let mut child = command.spawn()?;
+            let exit_status = child.wait().await?;
+            socket
+                .send(SuperExecResult {
+                    exit_code: exit_status.code().unwrap_or(127),
+                })
+                .await?;
+        }
+    }
+    Ok(())
+}
+
+async fn handle_escalate_session(socket: AsyncSocket) -> anyhow::Result<()> {
+    handle_escalate_session_with_decider(socket, decide_escalate).await
+}
+
+async fn escalate_task(socket: AsyncDatagramSocket) -> anyhow::Result<()> {
+    loop {
+        let (_, mut fds) = socket.receive_with_fds().await?;
+        if fds.len() != 1 {
+            tracing::error!("expected 1 fd in datagram handshake, got {}", fds.len());
+            continue;
+        }
+        let stream_socket = AsyncSocket::from_fd(fds.remove(0))?;
+        tokio::spawn(async move {
+            if let Err(err) = handle_escalate_session(stream_socket).await {
+                tracing::error!("escalate session failed: {err:?}");
+            }
+        });
+    }
+}
+
+fn get_bash_path() -> Result<String> {
+    std::env::var(BASH_PATH_ENV_VAR).context(format!("{BASH_PATH_ENV_VAR} must be set"))
+}
+
+async fn shell_exec(params: ExecParams) -> anyhow::Result<ExecResult> {
+    let bash_path = get_bash_path()?;
+    let ExecParams {
+        command,
+        workdir,
+        timeout_ms,
+    } = params;
+    let (escalate_server, escalate_client) = AsyncDatagramSocket::pair()?;
+    let client_socket = escalate_client.into_inner();
+    client_socket.set_cloexec(false)?;
+
+    let escalate_task = tokio::spawn(escalate_task(escalate_server));
+    let mut env = std::env::vars().collect::<HashMap<String, String>>();
+    env.insert(
+        ESCALATE_SOCKET_ENV_VAR.to_string(),
+        client_socket.as_raw_fd().to_string(),
+    );
+    env.insert(
+        "BASH_EXEC_WRAPPER".to_string(),
+        format!("{} escalate", std::env::current_exe()?.to_string_lossy()),
```

Comment:
use `shlex::join()`? or send as JSON and require the other side to parse it as JSON?

JSON approach eliminates `to_string_lossy()`?